File: pdu/javascript/hako_msgs/pdu_conv_DroneVisualStateArray.js
import * as PduUtils from '../pdu_utils.js';
import { DroneVisualStateArray } from './pdu_jstype_DroneVisualStateArray.js';
import { DroneVisualState } from './pdu_jstype_DroneVisualState.js';
import { binary_read_recursive_DroneVisualState, binary_write_recursive_DroneVisualState } from './pdu_conv_DroneVisualState.js';


/**
 * Deserializes a binary PDU into a DroneVisualStateArray object.
 * @param {ArrayBuffer} binary_data
 * @returns { DroneVisualStateArray }
 */
export function pduToJs_DroneVisualStateArray(binary_data) {
    const js_obj = new DroneVisualStateArray();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_DroneVisualStateArray(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_DroneVisualStateArray(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: sequence_id, type: uint32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 0, 4);
        js_obj.sequence_id = PduUtils.binToValue("uint32", bin);
    }
    
    // member: chunk_index, type: uint32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 4, 4);
        js_obj.chunk_index = PduUtils.binToValue("uint32", bin);
    }
    
    // member: chunk_count, type: uint32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 8, 4);
        js_obj.chunk_count = PduUtils.binToValue("uint32", bin);
    }
    
    // member: start_index, type: uint32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 12, 4);
        js_obj.start_index = PduUtils.binToValue("uint32", bin);
    }
    
    // member: valid_count, type: uint32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 16, 4);
        js_obj.valid_count = PduUtils.binToValue("uint32", bin);
    }
    
    // member: drones, type: hako_msgs/DroneVisualState (struct)

    { // varray
        const array_size = view.getInt32(base_off + 20, littleEndian);
        const offset_from_heap = view.getInt32(base_off + 20 + 4, littleEndian);
        const one_elm_size = 32;
        const array_value = [];
        for (let i = 0; i < array_size; i++) {
            const tmp_obj = new DroneVisualState();
            const item_offset = meta.heap_off + offset_from_heap + (i * one_elm_size);
            binary_read_recursive_DroneVisualState(meta, binary_data, tmp_obj, item_offset);
            array_value.push(tmp_obj);
        }
        js_obj.drones = array_value;
    }
    
    return js_obj;
}

/**
 * Serializes a DroneVisualStateArray object into a binary PDU.
 * @param { DroneVisualStateArray } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_DroneVisualStateArray(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_DroneVisualStateArray(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_DroneVisualStateArray(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: sequence_id, type: uint32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint32", js_obj.sequence_id, 4);
        allocator.add(bin, parent_off + 0);
    }
    
    // member: chunk_index, type: uint32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint32", js_obj.chunk_index, 4);
        allocator.add(bin, parent_off + 4);
    }
    
    // member: chunk_count, type: uint32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint32", js_obj.chunk_count, 4);
        allocator.add(bin, parent_off + 8);
    }
    
    // member: start_index, type: uint32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint32", js_obj.start_index, 4);
        allocator.add(bin, parent_off + 12);
    }
    
    // member: valid_count, type: uint32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint32", js_obj.valid_count, 4);
        allocator.add(bin, parent_off + 16);
    }
    
    // member: drones, type: hako_msgs/DroneVisualState (struct)

    { // varray
        const offset_from_heap = bw_container.heap_allocator.size();
        const array_size = js_obj.drones.length;
        const one_elm_size = 32;
        const array_base_offset = bw_container.heap_allocator.add(new ArrayBuffer(one_elm_size * array_size));
        for (let item_index = 0; item_index < array_size; item_index++) {
            const item_offset = array_base_offset + (item_index * one_elm_size);
            binary_write_recursive_DroneVisualState(item_offset, bw_container, bw_container.heap_allocator, js_obj.drones[item_index]);
        }

        const ref_buffer = new ArrayBuffer(8);
        const ref_view = new DataView(ref_buffer);
        ref_view.setInt32(0, array_size, littleEndian); // array_size
        ref_view.setInt32(4, offset_from_heap, littleEndian);
        allocator.add(ref_buffer, parent_off + 20);
    }
    
}

File: pdu/python/hako_msgs/pdu_conv_DroneVisualStateArray.py

import struct
from .pdu_pytype_DroneVisualStateArray import DroneVisualStateArray
from ..pdu_utils import *
from .. import binary_io

# dependencies for the generated Python class
from ..hako_msgs.pdu_conv_DroneVisualState import *



def pdu_to_py_DroneVisualStateArray(binary_data: bytearray) -> DroneVisualStateArray:
    py_obj = DroneVisualStateArray()
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        raise ValueError("Invalid PDU binary data: MetaData not found or corrupted")
    binary_read_recursive_DroneVisualStateArray(meta, binary_data, py_obj, binary_io.PduMetaData.PDU_META_DATA_SIZE)
    return py_obj


def binary_read_recursive_DroneVisualStateArray(meta: binary_io.PduMetaData, binary_data: bytearray, py_obj: DroneVisualStateArray, base_off: int):
    # array_type: single 
    # data_type: primitive 
    # member_name: sequence_id 
    # type_name: uint32 
    # offset: 0 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 0, 4)
    py_obj.sequence_id = binary_io.binTovalue("uint32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: chunk_index 
    # type_name: uint32 
    # offset: 4 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 4, 4)
    py_obj.chunk_index = binary_io.binTovalue("uint32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: chunk_count 
    # type_name: uint32 
    # offset: 8 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 8, 4)
    py_obj.chunk_count = binary_io.binTovalue("uint32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: start_index 
    # type_name: uint32 
    # offset: 12 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 12, 4)
    py_obj.start_index = binary_io.binTovalue("uint32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: valid_count 
    # type_name: uint32 
    # offset: 16 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 16, 4)
    py_obj.valid_count = binary_io.binTovalue("uint32", bin)
    
    # array_type: varray 
    # data_type: struct 
    # member_name: drones 
    # type_name: hako_msgs/DroneVisualState 
    # offset: 20 size: 32 
    # array_len: 8

    array_size = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, base_off + 20, 4))
    offset_from_heap = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, base_off + 20 + 4, 4))
    one_elm_size = 32
    i = 0
    array_value = []
    while i < array_size:
        tmp_py_obj = DroneVisualState()
        binary_read_recursive_DroneVisualState(meta, binary_data, tmp_py_obj, meta.heap_off + offset_from_heap + (i * one_elm_size))
        array_value.append(tmp_py_obj)
        i = i + 1
    py_obj.drones = array_value    
    
    return py_obj


def py_to_pdu_DroneVisualStateArray(py_obj: DroneVisualStateArray) -> bytearray:
    binary_data = bytearray()
    base_allocator = DynamicAllocator(False)
    bw_container = BinaryWriterContainer(binary_io.PduMetaData())
    binary_write_recursive_DroneVisualStateArray(0, bw_container, base_allocator, py_obj)

    # メタデータの設定
    total_size = base_allocator.size() + bw_container.heap_allocator.size() + binary_io.PduMetaData.PDU_META_DATA_SIZE
    bw_container.meta.total_size = total_size
    bw_container.meta.heap_off = binary_io.PduMetaData.PDU_META_DATA_SIZE + base_allocator.size()

    # binary_data のサイズを total_size に調整
    if len(binary_data) < total_size:
        binary_data.extend(bytearray(total_size - len(binary_data)))
    elif len(binary_data) > total_size:
        del binary_data[total_size:]

    # メタデータをバッファにコピー
    binary_io.writeBinary(binary_data, 0, bw_container.meta.to_bytes())

    # 基本データをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.base_off, base_allocator.to_array())

    # ヒープデータをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.heap_off, bw_container.heap_allocator.to_array())

    return binary_data

def binary_write_recursive_DroneVisualStateArray(parent_off: int, bw_container: BinaryWriterContainer, allocator, py_obj: DroneVisualStateArray):
    # array_type: single 
    # data_type: primitive 
    # member_name: sequence_id 
    # type_name: uint32 
    # offset: 0 size: 4 
    # array_len: 1
    type = "uint32"
    off = 0

    
    bin = binary_io.typeTobin(type, py_obj.sequence_id)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: chunk_index 
    # type_name: uint32 
    # offset: 4 size: 4 
    # array_len: 1
    type = "uint32"
    off = 4

    
    bin = binary_io.typeTobin(type, py_obj.chunk_index)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: chunk_count 
    # type_name: uint32 
    # offset: 8 size: 4 
    # array_len: 1
    type = "uint32"
    off = 8

    
    bin = binary_io.typeTobin(type, py_obj.chunk_count)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: start_index 
    # type_name: uint32 
    # offset: 12 size: 4 
    # array_len: 1
    type = "uint32"
    off = 12

    
    bin = binary_io.typeTobin(type, py_obj.start_index)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: valid_count 
    # type_name: uint32 
    # offset: 16 size: 4 
    # array_len: 1
    type = "uint32"
    off = 16

    
    bin = binary_io.typeTobin(type, py_obj.valid_count)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: varray 
    # data_type: struct 
    # member_name: drones 
    # type_name: hako_msgs/DroneVisualState 
    # offset: 20 size: 32 
    # array_len: 8
    type = "DroneVisualState"
    off = 20

    offset_from_heap = bw_container.heap_allocator.size()
    array_size = len(py_obj.drones)
    a_b = array_size.to_bytes(4, byteorder='little')
    o_b = offset_from_heap.to_bytes(4, byteorder='little')
    allocator.add(a_b + o_b, expected_offset=parent_off + off)
    bw_container.heap_allocator.add(bytearray(array_size * 32), expected_offset=offset_from_heap)
    for i, elm in enumerate(py_obj.drones):
        one_elm_size =  32
        binary_write_recursive_DroneVisualState((offset_from_heap + i * one_elm_size), bw_container, bw_container.heap_allocator, elm)
    

if __name__ == "__main__":
    import sys
    import json

    def print_usage():
        print(f"Usage: python -m pdu.python.pdu_conv_DroneVisualStateArray <read|write> [args...]")
        print(f"  read <input_binary_file> <output_json_file>")
        print(f"  write <input_json_file> <output_binary_file>")

    if len(sys.argv) < 2:
        print_usage()
        sys.exit(1)

    command = sys.argv[1]

    if command == "read":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)
        
        binary_filepath = sys.argv[2]
        output_json_filepath = sys.argv[3]

        with open(binary_filepath, "rb") as f:
            binary_data = bytearray(f.read())
        
        py_obj = pdu_to_py_DroneVisualStateArray(binary_data)
        
        with open(output_json_filepath, "w") as f:
            f.write(py_obj.to_json())

    elif command == "write":
        if len(sys.argv) != 4:
            print_usage()
            sys.exit(1)

        input_json_filepath = sys.argv[2]
        output_binary_filepath = sys.argv[3]

        with open(input_json_filepath, "r") as f:
            json_str = f.read()
        
        py_obj = DroneVisualStateArray.from_json(json_str)
        
        binary_data = py_to_pdu_DroneVisualStateArray(py_obj)

        with open(output_binary_filepath, "wb") as f:
            f.write(binary_data)

    else:
        print(f"Unknown command: {command}")
        print_usage()
        sys.exit(1)
